### PR TITLE
feature: dump localRegistry to cluster

### PR DESCRIPTION
**Pull request by Stevent-fei**
### Describe what this PR does / why we need it
#2056 
dump localRegistry to configMap

ha == true:
```shell
apiVersion: v1
data:
  registry: |
    external: null
    local:
      localregistry:
        registryconfig:
          domain: sea.hub
          port: 5000
          username: ""
          password: ""
        ha: true
        insecure: false
        cert:
          subjectaltname: null
      deployhosts:
      - 172.16.130.23
      - 172.16.130.24
      - 172.16.130.25
      vip: 10.103.97.2
kind: ConfigMap
metadata:
  creationTimestamp: "2023-04-10T02:43:55Z"
  name: sealer-registry
  namespace: kube-system
  resourceVersion: "758"
  uid: e7fc155e-cc73-4894-959b-a628e8c3f5e1
```

ha == false:
```shell
apiVersion: v1
data:
  registry: |
    external: null
    local:
      localregistry:
        registryconfig:
          domain: sea.hub
          port: 5000
          username: ""
          password: ""
        ha: false
        insecure: false
        cert:
          subjectaltname: null
      deployhosts:
      - 172.16.130.23
      vip: ""
kind: ConfigMap
metadata:
  creationTimestamp: "2023-04-10T01:52:24Z"
  name: sealer-registry
  namespace: kube-system
  resourceVersion: "734"
  uid: cb59317a-8f5e-4afc-9c84-863a981213d2
```

### CODE REVIEW

File: cmd/sealer/cmd/cluster/run.go [hunk 1]
```diff
@@ -333,8 +334,23 @@ func runClusterImage(imageEngine imageengine.Interface, cf clusterfile.Interface
...
+		ConfPath:        confPath,
+	}
+
+	registryInfo, err := json.Marshal(cluster.Spec.Registry.LocalRegistry)
+	if err != nil {
+		return err
+	}
+
+	// dump local registry to cluster
+	if err = clusterfile.SaveLocalRegistryToCluster(registryInfo, saveOptions.ConfPath); err != nil {
```
Comment (cubxxw) [inline cmd/sealer/cmd/cluster/run.go:348]: Could you write to save instead of run?

**Review by cubxxw — APPROVED** [review]

**Comment by kakaZhou719** [comment]
@Stevent-fei , i think we should save local registryConfig at install.go.  and only `v2.LocalRegistry` is not enough, should keep  registry VIP.

**Review by cubxxw — APPROVED** [review]

File: cmd/sealer/cmd/cluster/run.go [hunk 2]
```diff
@@ -307,14 +307,14 @@ func runClusterImage(imageEngine imageengine.Interface, cf clusterfile.Interface
 		return err
 	}
 
+	confPath := clusterruntime.GetClusterConfPath(imageSpec.ImageExtension.Labels)
```
Comment (kakaZhou719) [inline cmd/sealer/cmd/cluster/run.go:310]: not good, we use runtime driver to do it.

File: pkg/cluster-runtime/installer.go [hunk 3]
```diff
@@ -227,6 +228,11 @@ func (i *Installer) Install() error {
 		return err
 	}
 
+	// dump local registry info to cluster
+	if err = clusterfile.SaveLocalRegistryToCluster(clusterfile.SaveOptions{ConfPath: confPath}, i.regConfig.LocalRegistry, deployHosts, i.infraDriver); err != nil {
```
Comment (kakaZhou719) [inline pkg/cluster-runtime/installer.go:232]: could you set a func on Installer, not in clusterfile pkg to do it.

File: pkg/registry/local.go [hunk 4]
```diff
@@ -500,3 +495,30 @@ type DaemonLogOpts struct {
 	MaxFile       string `json:"max-file,omitempty"`
 	MaxSize       string `json:"max-size,omitempty"`
 }
+
+func NewLocalRegistryInfo(deployHosts []net.IP, localRegistry *v2.LocalRegistry, infraDriver infradriver.InfraDriver) *LocalRegistryInfo {
```
Comment (kakaZhou719) [inline pkg/registry/local.go:499]: add a new func looks like `GetRegistryInfo` . so we could get it after install

File: pkg/registry/local.go [hunk 5]
```diff
@@ -47,6 +48,12 @@ const (
 	LvscarePodFileName = "reg-lvscare.yaml"
 )
 
+type LocalRegistryInfo struct {
+	*v2.LocalRegistry
+	DeployHosts []net.IP
```
Comment (kakaZhou719) [inline pkg/registry/local.go:53]: add json tag `json:"deployHosts,omitempty"`

File: pkg/registry/local.go [hunk 6]
```diff
@@ -47,6 +48,12 @@ const (
...
+	Vip         string
```
Comment (kakaZhou719) [inline pkg/registry/local.go:54]: ditto

File: pkg/registry/external.go [hunk 7]
```diff
@@ -38,6 +38,14 @@ type externalConfigurator struct {
 	password             string
 }
 
+func (e *externalConfigurator) GetRegistryInfo(deployHosts []net.IP, LocalRegistry *v2.LocalRegistry) LocalRegistryInfo {
+	return LocalRegistryInfo{
+		LocalRegistry: LocalRegistry,
```
Comment (kakaZhou719) [inline pkg/registry/external.go:43]: why local ? this is externalConfigurator.

File: pkg/registry/configurator.go [hunk 8]
```diff
@@ -35,6 +33,8 @@ type Configurator interface {
 	UninstallFrom(masters, nodes []net.IP) error
 
 	GetDriver() (Driver, error)
+
+	GetRegistryInfo(deployHosts []net.IP, LocalRegistry *v2.LocalRegistry) LocalRegistryInfo
```
Comment (kakaZhou719) [inline pkg/registry/configurator.go:37]: GetRegistryInfo() RegistryInfo

File: pkg/cluster-runtime/installer.go [hunk 9]
```diff
@@ -227,6 +236,10 @@ func (i *Installer) Install() error {
 		return err
 	}
 
+	if err = i.setRegistryInfoToCluster(runtimeDriver, registryInfo); err != nil {
```
Comment (kakaZhou719) [inline pkg/cluster-runtime/installer.go:239]: saveRegistryInfo

File: pkg/registry/local.go [hunk 10]
```diff
@@ -500,3 +504,19 @@ type DaemonLogOpts struct {
 	MaxFile       string `json:"max-file,omitempty"`
 	MaxSize       string `json:"max-size,omitempty"`
 }
+
+func GetVIP(infraDriver infradriver.InfraDriver) string {
```
Comment (kakaZhou719) [inline pkg/registry/local.go:508]: GetRegistryVIP

File: pkg/registry/local.go [hunk 11]
```diff
@@ -55,6 +62,14 @@ type localConfigurator struct {
 	distributor          imagedistributor.Distributor
 }
 
+func (c *localConfigurator) GetRegistryInfo(deployHosts []net.IP, LocalRegistry *v2.LocalRegistry) LocalRegistryInfo {
+	return LocalRegistryInfo{
+		LocalRegistry: LocalRegistry,
+		DeployHosts:   deployHosts,
+		Vip:           GetVIP(c.infraDriver),
```
Comment (kakaZhou719) [inline pkg/registry/local.go:69]: vip is not required . so, if registry is not in ha mode, vip is meaningless .

File: pkg/registry/external.go [hunk 12]
```diff
@@ -38,6 +38,28 @@ type externalConfigurator struct {
...
+	)
+	if *clusterRegistry.LocalRegistry.HA {
+		deployHosts = masters
+		vip = GetRegistryVIP(e.infraDriver)
+	} else {
+		deployHosts = []net.IP{master0}
+	}
+
+	return RegistryInfo{
+		LocalRegistry: clusterRegistry.LocalRegistry,
```
Comment (kakaZhou719) [inline pkg/registry/external.go:57]: why local at external registry ?

File: pkg/registry/local.go [hunk 13]
```diff
@@ -47,6 +48,13 @@ const (
 	LvscarePodFileName = "reg-lvscare.yaml"
 )
 
+type RegistryInfo struct { // nolint
+	ExternalRegistry *v2.ExternalRegistry
+	LocalRegistry    *v2.LocalRegistry
+	DeployHosts      []net.IP `json:"deployHosts,omitempty"`
+	Vip              string   `json:"vip,omitempty"`
```
Comment (kakaZhou719) [inline pkg/registry/local.go:55]: vip and DeployHosts only valid in LocalRegistry. so could we add it to `LocalRegistry`

File: pkg/cluster-runtime/installer.go [hunk 14]
```diff
@@ -47,6 +51,9 @@ const (
 	CRILabel = "cluster.alpha.sealer.io/container-runtime-type"
 	// CRTLabel is key for get cluster runtime type.
 	CRTLabel = "cluster.alpha.sealer.io/cluster-runtime-type"
+
+	LocalRegistryConfigMapName     = "sealer-localregistry"
```
Comment (kakaZhou719) [inline pkg/cluster-runtime/installer.go:55]: if we save all registry info not only local, we can rename it `sealer-registry`

File: pkg/registry/local.go [hunk 15]
```diff
@@ -55,6 +67,21 @@ type localConfigurator struct {
 	distributor          imagedistributor.Distributor
 }
 
+func (c *localConfigurator) GetRegistryInfo() RegistryInfo {
+	localRegistryInfo := LocalRegistryInfo{}
+	localRegistryInfo.LocalRegistry = c.LocalRegistry
```
Comment (kakaZhou719) [inline pkg/registry/local.go:72]: merge two lines

File: pkg/registry/local.go [hunk 16]
```diff
@@ -47,6 +48,17 @@ const (
 	LvscarePodFileName = "reg-lvscare.yaml"
 )
 
+type LocalRegistryInfo struct {
+	LocalRegistry *v2.LocalRegistry
+	DeployHosts   []net.IP `json:"deployHosts,omitempty"`
+	Vip           string   `json:"vip,omitempty"`
+}
+
+type RegistryInfo struct { // nolint
```
Comment (kakaZhou719) [inline pkg/registry/local.go:57]: 1. move to configurator.go 
2. add json tag

File: pkg/registry/configurator.go [hunk 17]
```diff
@@ -18,14 +18,23 @@ import (
...
 )
 
+type LocalRegistryInfo struct {
+	LocalRegistry *v2.LocalRegistry
+	DeployHosts   []net.IP `json:"deployHosts,omitempty"`
+	Vip           string   `json:"vip,omitempty"`
+}
+
+type RegistryInfo struct { // nolint
+	ExternalRegistry *v2.ExternalRegistry `json:"external_registry"`
```
Comment (kakaZhou719) [inline pkg/registry/configurator.go:34]: `externalRegistry`,`localRegistry `. keep  code style same with others.

File: pkg/cluster-runtime/installer.go [hunk 18]
```diff
@@ -47,6 +51,9 @@ const (
 	CRILabel = "cluster.alpha.sealer.io/container-runtime-type"
 	// CRTLabel is key for get cluster runtime type.
 	CRTLabel = "cluster.alpha.sealer.io/cluster-runtime-type"
+
+	RegistryConfigMapName     = "sealer-registry"
+	RegistryConfigMapDataName = "registryInfo"
```
Comment (kakaZhou719) [inline pkg/cluster-runtime/installer.go:56]: registry

File: pkg/registry/configurator.go [hunk 19]
```diff
@@ -18,14 +18,23 @@ import (
...
 )
 
+type LocalRegistryInfo struct {
+	*v2.LocalRegistry
+	DeployHosts []net.IP `json:"deployHosts,omitempty"`
+	Vip         string   `json:"vip,omitempty"`
+}
+
+type RegistryInfo struct { // nolint
+	*v2.ExternalRegistry `json:"externalRegistry,omitempty"`
```
Comment (kakaZhou719) [inline pkg/registry/configurator.go:34]: external

File: pkg/registry/configurator.go [hunk 20]
```diff
@@ -18,14 +18,23 @@ import (
...
+	LocalRegistryInfo    `json:"localRegistry,omitempty"`
```
Comment (kakaZhou719) [inline pkg/registry/configurator.go:35]: local

**Pull request merged**